### PR TITLE
[Intl] Support ISO 3166-1 Alpha-3 country codes

### PR DESCRIPTION
**Pull request by terjebraten-certua**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no  
| Deprecations? | no
| Tests pass?   | yes 
| Fixed tickets | #20313
| Doc PR | symfony/symfony-docs#12105
| License       | MIT

This is a pre-requiste for fixing #20313 and only handles the changes to the INTL component.

### CODE REVIEW

**Comment by terjebraten-certua** [comment]
This a continuation of https://github.com/symfony/symfony/pull/32676

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php [hunk 1]
```diff
@@ -154,4 +177,33 @@ protected function generateRegionNames(ArrayAccessibleResourceBundle $localeBund
...
+                            $preferred = self::$preferredAlpha2ToAlpha3Mapping[$alpha2Code];
+                            // Only use the preferred mapping if it actually is in the mapping
+                            if ($alias === $preferred) {
+                                $alpha2ToAlpha3[$alpha2Code] = $preferred;
+                            }
+                        }
+                    }
+                }
+            }
+        }
```
Comment (derrabus) [inline src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php]: The nesting level of this method is pretty deep. You can flatten it by using early exits.

```php
foreach ($metadataBundle['alias']['territory'] as $alias => $data) {
    if (3 !== \strlen($alias) || 'overlong' !== $data['reason'] || ctype_digit($alias)) {
        continue;
    }

    $alpha2Code = $data['replacement'];
    if (!isset($alpha2Codes[$alpha2Code])) {
        continue;
    }
    // …
```

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php (hunk 1 above)
Comment (terjebraten-certua) [reply src/Symfony/Component/Intl/Data/Generator/RegionDataGenerator.php]: Done

**Comment by terjebraten-certua** [comment]
@derrabus Could you please complete your review?

**Comment by derrabus** [comment]
@terjebraten-certua I'm not familiar enough with the component to approve your changes, but I would trust that @ro0NL's review has been thorough. 😃 

Status: Reviewed

File: src/Symfony/Component/Intl/Countries.php [hunk 2]
```diff
@@ -52,6 +76,20 @@ public static function exists(string $country): bool
         }
     }
 
+    /**
+     * @param string $alpha3Code Alpha3 country code
+     */
```
Comment (fabpot) [inline src/Symfony/Component/Intl/Countries.php]: This phpdoc can be removed (we don't add them when they do not add any information already contained in the method signature).

File: src/Symfony/Component/Intl/Countries.php [hunk 3]
```diff
@@ -62,6 +100,16 @@ public static function getName(string $country, string $displayLocale = null): s
         return self::readEntry(['Names', $country], $displayLocale);
     }
 
+    /**
+     * Get country name from alpha3 code.
```
Comment (fabpot) [inline src/Symfony/Component/Intl/Countries.php]: Gets the country name from its alpha3 code.

File: src/Symfony/Component/Intl/Countries.php [hunk 4]
```diff
@@ -62,6 +100,16 @@ public static function getName(string $country, string $displayLocale = null): s
...
+     *
+     * @throws MissingResourceException if the country code does not exists
```
Comment (fabpot) [inline src/Symfony/Component/Intl/Countries.php]: exist

File: src/Symfony/Component/Intl/Languages.php [hunk 5]
```diff
@@ -79,6 +79,73 @@ public static function getAlpha3Code(string $language): string
...
+            self::getAlpha2Code($language);
+
+            return true;
+        } catch (MissingResourceException $e) {
+            return false;
+        }
+    }
+
+    /**
+     * Get language name from ISO 639-2 three-letter code.
```
Comment (fabpot) [inline src/Symfony/Component/Intl/Languages.php]: Gets

File: src/Symfony/Component/Intl/Languages.php [hunk 6]
```diff
@@ -79,6 +79,73 @@ public static function getAlpha3Code(string $language): string
...
+     *
+     * @throws MissingResourceException if the country code does not exists
+     */
+    public static function getAlpha3Name(string $language, string $displayLocale = null): string
+    {
+        return self::getName(self::getAlpha2Code($language), $displayLocale);
+    }
+
+    /**
+     * Get list of language names indexed with ISO 639-2 three-letter codes as keys.
```
Comment (fabpot) [inline src/Symfony/Component/Intl/Languages.php]: Gets the list ...

**Review by fabpot — APPROVED** [review]
with some minor CS comments

File: src/Symfony/Component/Intl/Countries.php (hunk 2 above)
Comment (terjebraten-certua) [reply src/Symfony/Component/Intl/Countries.php]: Done

File: src/Symfony/Component/Intl/Countries.php (hunk 3 above)
Comment (terjebraten-certua) [reply src/Symfony/Component/Intl/Countries.php]: Changed

File: src/Symfony/Component/Intl/Countries.php (hunk 4 above)
Comment (terjebraten-certua) [reply src/Symfony/Component/Intl/Countries.php]: Changed

**Comment by terjebraten-certua** [comment]
@fabpot I did all the changes you requested.

**Comment by fabpot** [comment]
Thank you @terjebraten-certua.

**Comment by TerjeBr** [comment]
@ro0NL Please have a look at #33136

**Pull request merged**